src/methods/extract.jl

```diff
@@ -47,66 +48,130 @@ collect(extract(st, pnts))
  (geometry = (0.52, 40.37), bio1 = missing, bio3 = missing, bio5 = missing, bio7 = missing, bio12 = missing)
  (geometry = (0.32, 40.24), bio1 = 16.321388f0, bio3 = 41.659454f0, bio5 = 30.029825f0, bio7 = 25.544561f0, bio12 = 480.0f0)
 ```
+
+Note: passing in arrays, geometry collections or feature collections 
+containing a mix of points and other geometries has undefined results.
 """
 function extract end
 function extract(x::RasterStackOrArray, data;
-    dims=DD.dims(x, DEFAULT_POINT_ORDER), kw...
+    dims=DD.dims(x, DEFAULT_POINT_ORDER), names=_names(x), geometry=true, index=false, kw...
 )
-    _extract(x, data; dims, names=_names(x), kw...)
+    
+    T = if geometry 
+        keys = index ? (:geometry, :index, names...,) : (:geometry, names...,)
+        NamedTuple{keys}
+    else
+        keys = index ? (:index, names...,) : (names...,)
+        NamedTuple{keys}
+    end
+    _extract(T, x, data; dims, names=NamedTuple{names}(names), kw...)
 end
-_extract(A::RasterStackOrArray, point::Missing; kw...) = missing
-function _extract(A::RasterStackOrArray, geom; kw...)
-    _extract(A, GI.geomtrait(geom), geom; kw...)
+
+function _extract(T, A::RasterStackOrArray, geom::Missing; names, kw...)
+    [_maybe_add_fields(T, A, map(_ -> missing, names), missing, missing)]
 end
-function _extract(A::RasterStackOrArray, ::Nothing, geoms; kw...)
-    geom1 = first(skipmissing(geoms))
-    if GI.isgeometry(geom1) || GI.isfeature(geom1) || GI.isfeaturecollection(geom1)
-        (_extract(A, g; kw...) for g in geoms)
+_extract(T, A::RasterStackOrArray, geom; kw...) = _extract(T, A, GI.geomtrait(geom), geom; kw...)
+_extract(T, A::RasterStackOrArray, ::Nothing, geom; kw...) = throw(ArgumentError("$geom is not a valid GeoInterface.jl geometry"))
+function _extract(T, A::RasterStackOrArray, ::Nothing, geoms::AbstractArray; names, skipmissing=false, kw...)
+    # Handle empty / all missing cases
+    (length(geoms) > 0 && any(!ismissing, geoms)) || return T[]
+    geom1 = first(Base.skipmissing(geoms)) # TODO: will fail if `geoms` is empty or all missing
```

Comment:
As is `extract(araster, [(valid_geom, invalid_geom)])` and `extract(araster, [(invalid_geom, valid_geom)])` return different errors. Explicitly checking for invalid geometries would solve this.

```suggestion
    # Handle cases with some invalid geometries
    invalid_geom_idx = findfirst(g -> !ismissing(g) && GI.geomtrait(g) === nothing, geoms)
    invalid_geom_idx === nothing || throw(ArgumentError("$(geoms[invalid_geom_idx]) is not a valid GeoInterface.jl geometry"))

    geom1 = first(Base.skipmissing(geoms))
```